Allow build and test on macosx with non default SDK

fuhlig1 opened the pull request:
The two commits allow to build and test FairRoot on macosx with a non default SDK. To use this feature a upcoming version of FairSoft is needed. Currently the head of jan24_patches which will soon get the tag jan24p6 or the release candidate branch for the upcoming new release sep25 will be okay.

For Linux systems or macosx systems using the default SDK the commits are not needed.

The two commits must be ported to the following branches:

- v18.6_patches
- v18.8_patches
- v19.0_patches

---

Checklist:

* [x ] Followed the [Contributing Guidelines](https://github.com/FairRootGroup/FairRoot/blob/dev/CONTRIBUTING.md)

fuhlig1 committed:
fix(build): Properly pass CMAKE_OSX_SYSROOT to rootcint

This change allows to build dictionaries with a non default macosx
software development package (SDK). This is needed to be able to
compile older ROOT versions on new compiler versions.

fuhlig1 commented on cmake/scripts/generate_dictionary_root.sh.in:
```diff
@@ -10,4 +10,8 @@
 
+if [[ -n @CMAKE_OSX_SYSROOT@ ]]; then
+  export SDKROOT=@CMAKE_OSX_SYSROOT@
+fi
```
Done

fuhlig1 commented on templates/test_project_root_containers.sh.in:
```diff
@@ -12,18 +12,26 @@ cd project_root_containers
 
+if [[ -n @CMAKE_OSX_SYSROOT@ ]]; then
+  parameters="$parameters -DCMAKE_OSX_SYSROOT=@CMAKE_OSX_SYSROOT@"
+fi
```
Done

fuhlig1 commented on templates/test_project_stl_containers.sh.in:
```diff
@@ -14,16 +14,24 @@ mkdir build
 
+if [[ -n @CMAKE_OSX_SYSROOT@ ]]; then
+  parameters="$parameters -DCMAKE_OSX_SYSROOT=@CMAKE_OSX_SYSROOT@"
+fi
```
Done

fuhlig1 committed:
fix(tests): Fix template tests for macosx

Pass the needed information about CMAKE_OSX_SYSROOT to rootcling
such that the template tests also work on macosx with a non standard
SDK.

fuhlig1 commented:
@dennisklein,

could you please have a look if the commits are okay from your side. I would like to merge them such that I can port them to the patches branches. My tests run nearly on all systems. The only problem still is the on ubuntu 24.04 where doing a 'cp -a' of the templates directory results in a crash of the tests but this problem is unrelated to the actual test performed.

fuhlig1 commented:
@dennisklein,

the same problem with `cp -a` to the /tmp directory is present for OpenSuse. `cp -a` to /scratch works on both systems. Looking more closely in the problem the preservation of the ownership is problematic. 

So the problem is probably due to the handling of the /tmp directory from the container.

A possible workaround would be to use `cp -dR` which is equivalent to `cp -a` except of the explicit preservation of the permissions. Checking the files in the /tmp directory the timestamp, the access rights as well as the ownership look okay. For me it looks like that `cp -dR` does the permission preservation implicitly.

dennisklein committed:
refactor(templates): Consolidate duplicate test project scripts

Replaces test_project_root_containers.sh.in and test_project_stl_containers.sh.in
with a single parameterized test_project.sh.in script to eliminate code duplication.

dennisklein committed:
refactor(templates): Improve test_project.sh robustness

- Add proper argument parsing with case statement
- Use modern cmake -S/-B and --build syntax
- Replace backticks with $() syntax
- Add trap for cleanup on exit
- Use bash arrays for cmake parameters
- Add -u and -o pipefail to set flags
- Make -x conditional on DEBUG environment variable
- Add template directory validation
- Use readonly for immutable variables
- Redirect errors to stderr
- Add --help flag

dennisklein commented:
> @dennisklein,
> 
> the same problem with `cp -a` to the /tmp directory is present for OpenSuse. `cp -a` to /scratch works on both systems. Looking more closely in the problem the preservation of the ownership is problematic.
> 
> So the problem is probably due to the handling of the /tmp directory from the container.
> 
> A possible workaround would be to use `cp -dR` which is equivalent to `cp -a` except of the explicit preservation of the permissions. Checking the files in the /tmp directory the timestamp, the access rights as well as the ownership look okay. For me it looks like that `cp -dR` does the permission preservation implicitly.

A simple `cp -r` will also do the job here, no? No objections on changing the `cp` parameters.

dennisklein reviewed (dismissed):
looks fine to me, I also did some deduplication/refactorings on the scripts, see https://github.com/fuhlig1/FairRoot/compare/vdev_patches...dennisklein:FairRoot:vdev_patches

If you like them, feel free to fetch them :)

dennisklein commented on templates/test_project_root_containers.sh.in:
```diff
@@ -14,14 +14,20 @@ mkdir build
+  echo "*** Calling cmake configure again:"
+  # Check if all the cache variables
+  # are good for a reconfiguration step
+  cmake ..
```
it should be cached from the first run

fuhlig1 committed:
Fix copy error on some systems

Using `cp -a` mode to the tmp directory from a container fails for
ubuntu 24.04 and opensuse 16.0 due failure of preserving permissions.
Using `cp -RP` will copy the directory recursivly not following any
symbolic links.

fuhlig1 committed:
fix(tests): Fix template tests

Shell script must be executable.

fuhlig1 commented:
> looks fine to me, I also did some deduplication/refactorings on the scripts, see [fuhlig1/FairRoot@vdev_patches...dennisklein:FairRoot:vdev_patches](https://github.com/fuhlig1/FairRoot/compare/vdev_patches...dennisklein:FairRoot:vdev_patches)
> 
> If you like them, feel free to fetch them :)

Have you tested the changes? I had first a problem that the shell script for the template test wasn't executable which was easy to fix but now I run into problems that headers are not found during dictionary creation.

dennisklein commented:
Nope, had no fairsoft installation lying around.

dennisklein commented:
Compiling FairSoft now and going to test the script changes

fuhlig1 commented:
I found the first issue but at least for v18.6_patches I wonder how the compilation could have worked before.

dennisklein commented:
Tested FairRoot `(this PR branch)` + FairSoft `jan24p5` installed in `/cvmfs/fairsoft_dev.gsi.de/ci/for-fairsoft/latest/container/ubuntu.22.04.legacy.sif`. Besides the missing execution bit, this worked (Note: I built and installed FairRoot before running the tests):
```
Singularity> ctest --test-dir build -R container --output-on-failure -j16
Internal ctest changing into directory: /home/dklein/projects/FairSoft/FairRoot/build
Test project /home/dklein/projects/FairSoft/FairRoot/build
    Start 5: project_stl_containers_double
    Start 3: project_root_containers_double
    Start 4: project_stl_containers
    Start 2: project_root_containers
1/4 Test #2: project_root_containers ..........   Passed   25.90 sec
2/4 Test #4: project_stl_containers ...........   Passed   25.91 sec
3/4 Test #5: project_stl_containers_double ....   Passed   26.17 sec
4/4 Test #3: project_root_containers_double ...   Passed   26.28 sec

100% tests passed, 0 tests failed out of 4

Total Test time (real) =  26.28 sec
```

Also ran the full test suite
```
100% tests passed, 0 tests failed out of 68

Total Test time (real) =  39.08 sec
```

Let me know, if I can help with any different environment.

fuhlig1 commented:
Ok. I don't understand why it worked in your case but at least in the debian13 container the execution of the generated script test_project.sh haven't worked. It simply had no execution rights. After I changed that the tests for this branch, the v19.0_patches and the v18.8_patches branches worked like expected.
Unfortunately I started with the v18.6_patches branch and there beside the problem with the execution rights there were some errors in the build system as well as the code. This couldn't have worked before but when running the FairSoft test suite the test for v18.6_patches always worked.
I will do the tests again and if everything now works I will merge.

dennisklein commented:
> It simply had no execution rights.

Yes, this was missing, as I acknowledged in my post ;)
